server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletGoalState.java

```diff
@@ -145,6 +146,10 @@ public String getResourceGroup() {
         }
       }
 
+      if (VolumeUtil.needsVolumeReplacement(params.getVolumeReplacements(), tm)) {
```

Comment:
The reason I called it twice was because I think having the goal state determined in a single place in the code makes it easier to review and understand.  The goal state used to be determined in a multiple places in the code (spread between manager and TGW code) and it was harder to understand and reason about the goal state (needed to know all the places in the code that determined the goal state).  Would like to keep it in a single place for this reason.  Once the manager has done its first pass the volume checks will always be super quick because the list of replacements is empty, so its will not matter that its called twice then.  Also needsVolumeReplacement method was optimized to do less allocation, probably making it faster.

Reply:
We could possibly pass a `boolean needsVolumeReplacement` to the computeGoalState function to avoid computing twice if its actually a performance issue.

Reply:
Also the goal state needs to be properly computed in the TabletMgmtIter so that it will return an action of needs location update in addition to returning needs volume replacement.  Need to get tablets unassigned before doing the volume update.

Reply:
I think the change I made in b482623 alleviates my concern - guarding the check for volume replacement depending on the size of the volume replacements list. This list is empty (see createTabletManagementParameters) when no more volume replacements are occurring in the TGW. I also added some protection to make sure that the tablet is unassigned when doing the volume replacement.